mip/driver_imx_rt1020.c

```diff
@@ -0,0 +1,302 @@
+
+  (void) data;
+  return true;
+}
+
+// Transmit frame
+static uint32_t s_txno;
+
+static size_t mip_driver_imx_rt1020_tx(const void *buf, size_t len, void *userdata) {
+
+  if (len > sizeof(tx_data_buffer[ENET_TXBD_NUM])) {
+  //  MG_ERROR(("Frame too big, %ld", (long) len));
+    len = 0;  // Frame is too big
+  } else if ((tx_buffer_descriptor[s_txno].control & BIT_SET(15))) {
+  MG_ERROR(("No free descriptors"));
+    // printf("D0 %lx SR %lx\n", (long) s_txdesc[0][0], (long) ETH->DMASR);
+    len = 0;  // All descriptors are busy, fail
+  } else {
+    memcpy(tx_data_buffer[s_txno], buf, len);     // Copy data
+    tx_buffer_descriptor[s_txno].length = (uint16_t) len;  // Set data len
+    tx_buffer_descriptor[s_txno].control |= (uint16_t)(BIT_SET(10)); // TC (transmit CRC)
+    //  tx_buffer_descriptor[s_txno].control &= (uint16_t)(BIT_SET(14) | BIT_SET(12)); // Own doesn't affect HW
+    tx_buffer_descriptor[s_txno].control |= (uint16_t)(BIT_SET(15) | BIT_SET(11)); // R+L (ready+last)
+    ENET->TDAR = BIT_SET(24); // Descriptor updated. Hand over to DMA.
+    // INFO
+    // Relevant Descriptor bits: 15(R)  Ready
+    //                           11(L)  last in frame
+    //                           10(TC) transmis CRC
+    // __DSB(); // ARM errata 838869 Cortex-M4, M4F, M7, M7F: "store immediate overlapping
+                // exception" return might vector to incorrect interrupt.
+    if (++s_txno >= ENET_TXBD_NUM) s_txno = 0;
+  }
+  (void) userdata;
+  return len;
+}
+
+// IRQ (RX)
+static uint32_t s_rxno;
+
+void ENET_IRQHandler(void) {
+  ENET->EIMR = 0;           // Mask interrupts.
+  uint32_t eir = ENET->EIR; // Read EIR
+  ENET->EIR = 0xffffffff;   // Clear interrupts
+
+  if (eir & EIMR_RX_ERR) // Global mask used
+  {
+    if (rx_buffer_descriptor[s_rxno].control & BIT_SET(15)) {
+      ENET->EIMR = EIMR_RX_ERR; // Enable interrupts
+      return;  // Empty? -> exit.
+    }
+    // Read inframes
+    else { // Frame received, loop
+      for (uint32_t i = 0; i < 10; i++) {  // read as they arrive but not forever
+        if (rx_buffer_descriptor[s_rxno].control & BIT_SET(15)) break;  // exit when done
+        uint32_t len = (rx_buffer_descriptor[s_rxno].length);
+        if (s_rx != NULL)
+          s_rx(rx_buffer_descriptor[s_rxno].buffer, len > 4 ? len - 4 : len, s_rxdata);
+        rx_buffer_descriptor[s_rxno].control |= BIT_SET(15); // Inform DMA RX is empty
+        if (++s_rxno >= ENET_RXBD_NUM) s_rxno = 0;
+      }
+    }
+  }
+  ENET->EIMR = EIMR_RX_ERR; // Enable interrupts
+}
+
+// Up/down status
+static bool mip_driver_imx_rt1020_up(void *userdata) {
+  uint32_t bsr = eth_read_phy(PHY_ADDR, PHY_BSR);
+  (void) userdata;
+  return bsr & BIT_SET(2) ? 1 : 0;
+}
+
+// Set receive callback for interrupt-driven drivers
+static void mip_driver_imx_rt1020_setrx(void (*rx)(void *, size_t, void *),
+                                   void *rxdata) {
+  s_rx = rx;
+  s_rxdata = rxdata;
+}
```

Comment:
Oh, this is really old stuff... time passes quickly (and chances are few)...

Reply:
Right i posted code from branch, let me work on update.

Reply:
BTW, that is a quotation from a Toto song... I should have changed it to "and changes aren't few"